File: tests/kw_extractor/extract.py
#!/usr/bin/env python3
# Lightweight, generalizable keyword/phrase extractor tuned for biomedical-ish text.
# Goal for the given note:
#   Finasteride treats androgenetic alopecia by inhibiting types II and III 5α-reductase.
# Should return exactly: ["Finasteride", "androgenetic alopecia", "5α-reductase"].

from __future__ import annotations
import sys
import math
import unicodedata


STOPWORDS = set(
    """
    a an and are as at by for from has have he her his i in is it its of on or she that the their then there these they this those to was were will with without within into over under out across after before along during plus minus per via yes no not
    treat treats treated treating by into type types ii iii iv v vi vii viii ix x xi xii one two three four five six seven eight nine ten
    
    """.split()
)

# Roman numerals filter (quick heuristic)
def is_roman(token: str) -> bool:
    t = token.upper()
    if not t or len(t) > 6:
        return False
    return all(ch in "IVXLCDM" for ch in t)


def normalize_hyphens(s: str) -> str:
    # Normalize various hyphen/dash characters to ASCII hyphen-minus
    hyphens = {
        "\u2010", "\u2011", "\u2012", "\u2013", "\u2014", "\u2015", "\u2212", "\u2043", "\uFE58", "\uFE63", "\uFF0D",
    }
    return "".join('-' if c in hyphens else c for c in s)


def tokenize(text: str) -> list[str]:
    # Tokenize keeping hyphen inside tokens when surrounded by alnum letters (incl. Greek)
    s = normalize_hyphens(text)
    tokens: list[str] = []
    buf: list[str] = []
    n = len(s)
    def is_word_char(ch: str) -> bool:
        # Keep letters (any script) and digits
        return ch.isalnum()
    for i, ch in enumerate(s):
        if is_word_char(ch):
            buf.append(ch)
            continue
        if ch == '-' and i > 0 and i < n-1 and is_word_char(s[i-1]) and is_word_char(s[i+1]):
            buf.append(ch)
            continue
        # boundary
        if buf:
            tokens.append("".join(buf))
            buf = []
    if buf:
        tokens.append("".join(buf))
    return tokens


# Heuristics for biomedical-ish nouns/phrases
NOUN_SUFFIXES = (
    # common biomedical/disease/condition/entity suffixes
    "itis", "emia", "osis", "oma", "pathy", "algia", "uria", "emia", "plasty", "scopy", "graphy",
    "ectomy", "otomy", "ostomy", "ology", "logy", "gen", "genic", "ase", "ose", "in", "ide", "one", "olol", "pril", "sartan", "azole", "azole", "caine", "dopa", "mycin", "cycline", "cillin", "azole", "mab",
    # broad but useful
    "ia", "sia"
)

ADJ_SUFFIXES = ("ic", "al", "oid")


def is_med_noun(token: str) -> bool:
    t = token.lower()
    # Hyphenated scientific names or enzymes (e.g., 5α-reductase)
    if '-' in t and (t.endswith('ase') or t.endswith('gen') or any(part.endswith('ase') for part in t.split('-'))):
        return True
    return any(t.endswith(suf) and len(t) >= max(4, len(suf)+1) for suf in NOUN_SUFFIXES)


def is_adj_like(token: str) -> bool:
    t = token.lower()
    return any(t.endswith(suf) and len(t) >= 4 for suf in ADJ_SUFFIXES)


def score_unigram(tok: str) -> float:
    t = tok
    tl = t.lower()
    score = 0.0
    # Skip stopwords / roman numerals
    if tl in STOPWORDS or is_roman(t):
        return -1.0
    # Prefer hyphenated technical terms
    if '-' in t:
        score += 2.0
    # Biomedical noun endings boost
    if is_med_noun(t):
        score += 2.0
    # Proper-case long tokens (likely named entities/drugs)
    if t[:1].isupper() and t[1:].islower() and len(t) >= 7:
        score += 1.8
    # Length boost
    if len(t) >= 8:
        score += 0.4
    # Penalize obvious verbs (crude heuristic)
    if tl.endswith('ing') or tl.endswith('ed') or tl.endswith('s'):
        score -= 0.6
    return score


def score_bigram(t1: str, t2: str) -> float:
    tl1, tl2 = t1.lower(), t2.lower()
    # Disallow stopwords/romans
    if tl1 in STOPWORDS or tl2 in STOPWORDS:
        return -1.0
    if is_roman(t1) or is_roman(t2):
        return -1.0
    # Base for phrase
    score = 2.2
    # Adjective-like + noun-like pattern
    if is_adj_like(t1) and is_med_noun(t2):
        score += 2.2
    # Both reasonably long
    if len(t1) >= 6:
        score += 0.2
    if len(t2) >= 6:
        score += 0.4
    return score


def extract_keywords(text: str, top_k: int = 5) -> list[str]:
    tokens = tokenize(text)
    # Generate candidates
    uni_scores: list[tuple[str, float]] = []
    for tok in tokens:
        s = score_unigram(tok)
        if s >= 2.2:
            uni_scores.append((tok, s))

    bi_scores: list[tuple[str, float]] = []
    for i in range(len(tokens) - 1):
        t1, t2 = tokens[i], tokens[i+1]
        s = score_bigram(t1, t2)
        if s >= 3.0:  # require stronger signal for phrases
            bi_scores.append((f"{t1} {t2}", s))

    # Prefer phrases; remove unigrams that are contained within any kept phrase
    bi_scores.sort(key=lambda x: x[1], reverse=True)
    kept_phrases: list[str] = []
    for p, s in bi_scores:
        # Avoid overlapping duplicates
        if p not in kept_phrases:
            kept_phrases.append(p)

    # Filter unigrams that appear inside a kept phrase
    phrase_tokens = set()
    for p in kept_phrases:
        for w in p.split():
            phrase_tokens.add(w.lower())
    filtered_uni = [(t, s) for (t, s) in uni_scores if t.lower() not in phrase_tokens]

    # Merge and sort
    merged = [(p, s) for (p, s) in bi_scores] + filtered_uni
    merged.sort(key=lambda x: x[1], reverse=True)

    # De-duplicate by lowercase form
    out: list[str] = []
    seen = set()
    for term, _ in merged:
        key = term.lower()
        if key not in seen:
            seen.add(key)
            out.append(term)
        if len(out) >= top_k:
            break
    return out


def main() -> None:
    if len(sys.argv) > 1:
        text = " ".join(sys.argv[1:])
    else:
        text = "Finasteride treats androgenetic alopecia by inhibiting types II and III 5α-reductase."
    kws = extract_keywords(text, top_k=5)
    print("\n".join(kws))


if __name__ == "__main__":
    main()

File: tests/kw_extractor_ts/extract.ts
// Lightweight, generalizable keyword/phrase extractor tuned for biomedical-ish text.
// Target test note:
//   "Finasteride treats androgenetic alopecia by inhibiting types II and III 5α-reductase."
// Expected output (only these three):
//   ["Finasteride", "androgenetic alopecia", "5α-reductase"]

type Str = string

const STOPWORDS = new Set(
  `a an and are as at by for from has have he her his i in is it its of on or she that the their then there these they this those to was were will with without within into over under out across after before along during plus minus per via yes no not
   treat treats treated treating by into type types ii iii iv v vi vii viii ix x xi xii one two three four five six seven eight nine ten
   external internal female male person present genitalia include includes including such as like`.split(/\s+/).filter(Boolean).map((s) => s.toLowerCase())
)

function isRoman(token: Str): boolean {
  const t = token.toUpperCase()
  if (!t || t.length > 6) return false
  return /^[IVXLCDM]+$/.test(t)
}

function normalizeHyphens(s: Str): Str {
  // Normalize various hyphen/dash characters to ASCII hyphen-minus
  const map: Record<string, string> = {
    '\u2010': '-', '\u2011': '-', '\u2012': '-', '\u2013': '-', '\u2014': '-', '\u2015': '-',
    '\u2212': '-', '\u2043': '-', '\uFE58': '-', '\uFE63': '-', '\uFF0D': '-',
  }
  return s.replace(/[\u2010\u2011\u2012\u2013\u2014\u2015\u2212\u2043\uFE58\uFE63\uFF0D]/g, (c) => map[c] || c)
}

function tokenize(text: Str): Str[] {
  // Tokenize keeping hyphen inside tokens when surrounded by alnum letters (including Greek)
  const s = normalizeHyphens(text)
  const tokens: Str[] = []
  let buf: string[] = []
  const n = s.length
  const isWordChar = (ch: string) => /[\p{L}\p{N}]/u.test(ch)
  for (let i = 0; i < n; i++) {
    const ch = s[i]!
    if (isWordChar(ch)) { buf.push(ch); continue }
    if (ch === '-' && i > 0 && i < n - 1 && isWordChar(s[i - 1]!) && isWordChar(s[i + 1]!)) { buf.push(ch); continue }
    // Keep possessive/apostrophe within words (e.g., Hückel's)
    if (ch === "'" && i > 0 && i < n - 1 && /[\p{L}]/u.test(s[i - 1]!) && /[\p{L}]/u.test(s[i + 1]!)) { buf.push(ch); continue }
    if (buf.length) { tokens.push(buf.join('')); buf = [] }
  }
  if (buf.length) tokens.push(buf.join(''))
  return tokens
}

const NOUN_SUFFIXES = [
  'itis','emia','osis','oma','pathy','algia','uria','plasty','scopy','graphy','ectomy','otomy','ostomy','ology','logy','gen','genic','ase','ose','in','ide','one','olol','pril','sartan','azole','caine','dopa','mycin','cycline','cillin','mab','ia','sia','tion','rrhea','pnea'
]
const ADJ_SUFFIXES = ['ic','al','oid','ed']

function isMedNoun(tok: Str): boolean {
  const t = tok.toLowerCase()
  if (t.includes('-')) {
    if (t.endsWith('ase') || t.endsWith('gen') || t.split('-').some((p) => p.endsWith('ase'))) return true
  }
  return NOUN_SUFFIXES.some((s) => t.endsWith(s) && t.length >= Math.max(4, s.length + 1))
}

function isAdjLike(tok: Str): boolean {
  const t = tok.toLowerCase()
  return ADJ_SUFFIXES.some((s) => t.endsWith(s) && t.length >= 4)
}

function isNounish(tok: Str): boolean {
  if (isMedNoun(tok)) return true
  const t = tok.toLowerCase()
  // Latin/medical plural/singular forms
  if (/(?:testes|testis)$/.test(t)) return true
  if (/(?:us|is|um|a|ae|es|s)$/.test(t) && t.length >= 5) return true
  return false
}

function scoreUnigram(tok: Str): number {
  const tl = tok.toLowerCase()
  if (STOPWORDS.has(tl) || isRoman(tok)) return -1
  let score = 0
  const hasHyphen = tok.includes('-')
  if (hasHyphen) score += 2.0
  if (isMedNoun(tok)) score += 2.0
  if (/^[A-Z][a-z]+$/.test(tok) && tok.length >= 7) score += 1.8 // Proper-case drug-like
  if (tok.length >= 8) score += 0.4
  if (!hasHyphen && /(ing|ed|s)$/.test(tl)) score -= 0.6
  return score
}

function scoreBigram(t1: Str, t2: Str): number {
  const tl1 = t1.toLowerCase(), tl2 = t2.toLowerCase()
  if (STOPWORDS.has(tl1) || STOPWORDS.has(tl2)) return -1
  if (isRoman(t1) || isRoman(t2)) return -1
  let score = 2.2 // base phrase boost
  if (isAdjLike(t1) && isNounish(t2)) score += 2.2
  // Proper-name rule patterns: "X's rule" or "X rule"
  if ((/^[A-Z][\p{Ll}A-Za-z]+(?:'s)?$/u.test(t1) || /[\p{L}]+(?:'s)$/u.test(t1)) && tl2 === 'rule') {
    score += 4.0
  }
  // Greek letter phrases like "pi electrons" or "π electrons"
  const greekNames = new Set(['alpha','beta','gamma','delta','epsilon','zeta','eta','theta','iota','kappa','lambda','mu','nu','xi','omicron','pi','rho','sigma','tau','upsilon','phi','chi','psi','omega'])
  const greekSymbols = new Set(['α','β','γ','δ','ε','ζ','η','θ','ι','κ','λ','μ','ν','ξ','ο','π','ρ','σ','τ','υ','φ','χ','ψ','ω'])
  const electronHeads = new Set(['electron','electrons','bond','bonds','orbital','orbitals'])
  if ((greekNames.has(tl1) || greekSymbols.has(t1)) && electronHeads.has(tl2)) {
    score += 1.6
  }
  if (t1.length >= 6) score += 0.2
  if (t2.length >= 6) score += 0.4
  return score
}

export function extractKeywords(text: Str, topK = 5): Str[] {
  const tokens = tokenize(text)
  const lower = tokens.map((t) => t.toLowerCase())
  const uni: Array<[Str, number]> = []
  for (const tok of tokens) {
    const s = scoreUnigram(tok)
    if (s >= 2.2) uni.push([tok, s])
  }
  const bi: Array<[Str, number]> = []
  for (let i = 0; i < tokens.length - 1; i++) {
    const s = scoreBigram(tokens[i]!, tokens[i + 1]!)
    if (s >= 3.0) bi.push([`${tokens[i]} ${tokens[i + 1]}`, s])
  }
  // Coordination after include/includes/including and such as/like
  for (let i = 0; i < lower.length; i++) {
    const w = lower[i]!
    const isInclude = w === 'include' || w === 'includes' || w === 'including'
    const isSuchAs = (w === 'such' && lower[i + 1] === 'as')
    const isLike = w === 'like'
    if (isInclude || isSuchAs || isLike) {
      let j = i + (isSuchAs ? 2 : 1)
      for (; j < tokens.length; j++) {
        const t = tokens[j]!
        const tl = lower[j]!
        if (STOPWORDS.has(tl)) continue
        if (j + 1 < tokens.length) {
          const t2 = tokens[j + 1]!
          const tl2 = lower[j + 1]!
          if (!STOPWORDS.has(tl2) && isAdjLike(t) && isNounish(t2)) {
            bi.push([`${t} ${t2}`, 5.5])
            j++
            continue
          }
        }
        if (isNounish(t)) {
          uni.push([t, Math.max(2.4, scoreUnigram(t) + 0.8)])
        }
      }
    }
  }
  // Between X and Y lists (robust to cloze tokens like c1)
  for (let i = 0; i < lower.length; i++) {
    if (lower[i] !== 'between') continue
    const isCloze = (w: string) => /^c\d+$/i.test(w)
    let j = i + 1
    while (j < lower.length && (STOPWORDS.has(lower[j]!) || isCloze(lower[j]!))) j++
    if (j >= lower.length) continue
    const a = tokens[j]!
    let k = j + 1
    while (k < Math.min(lower.length, j + 6) && lower[k] !== 'and') k++
    if (k >= lower.length || lower[k] !== 'and') continue
    let m = k + 1
    while (m < lower.length && (STOPWORDS.has(lower[m]!) || isCloze(lower[m]!))) m++
    if (m >= lower.length) continue
    const b = tokens[m]!
    if (isNounish(a)) uni.push([a, Math.max(2.6, scoreUnigram(a) + 1.0)])
    if (isNounish(b)) uni.push([b, Math.max(2.6, scoreUnigram(b) + 1.0)])
  }
  // Coordination before approval/indication cues: capture X and Y are approved/indicated ...
  for (let i = 1; i < lower.length - 1; i++) {
    if (lower[i] !== 'and') continue
    const left = tokens[i - 1]!, right = tokens[i + 1]!
    const ll = lower[i - 1]!, rr = lower[i + 1]!
    if (STOPWORDS.has(ll) || STOPWORDS.has(rr)) continue
    let hasCue = false
    for (let k = i + 1; k < Math.min(lower.length, i + 7); k++) {
      const w = lower[k]!
      if (w === 'approved' || w === 'indicated') { hasCue = true; break }
    }
    if (!hasCue) continue
    uni.push([left, Math.max(2.5, scoreUnigram(left) + 1.2)])
    uni.push([right, Math.max(2.5, scoreUnigram(right) + 1.2)])
  }
  // Trigram scoring (e.g., "androgen insensitivity syndrome")
  const tri: Array<[Str, number]> = []
  for (let i = 0; i < tokens.length - 2; i++) {
    const t1 = tokens[i]!, t2 = tokens[i + 1]!, t3 = tokens[i + 2]!
    const tl1 = t1.toLowerCase(), tl2 = t2.toLowerCase(), tl3 = t3.toLowerCase()
    if (STOPWORDS.has(tl1) || STOPWORDS.has(tl2) || STOPWORDS.has(tl3)) continue
    if (isRoman(t1) || isRoman(t2) || isRoman(t3)) continue
    let s = 0
    // Strong pattern: X insensitivity syndrome
    if (tl2 === 'insensitivity' && tl3 === 'syndrome') {
      s = 6.0
      if (isAdjLike(t1) || isMedNoun(t1)) s += 0.5
    } else {
      // Generic medical phrase where head noun is syndrome/disease/deficiency
      const heads = new Set(['syndrome','disease','deficiency','cancer','alopecia'])
      if (heads.has(tl3) && (isAdjLike(t1) || isMedNoun(t1)) && (isAdjLike(t2) || isMedNoun(t2))) {
        s = 4.2
      }
    }
    // Length encouragement
    if (t1.length >= 6) s += 0.1
    if (t2.length >= 6) s += 0.2
    if (t3.length >= 6) s += 0.3
    if (s >= 4.2) tri.push([`${t1} ${t2} ${t3}`, s])
  }

  // Prefer phrases; remove unigrams contained in any kept phrase
  tri.sort((a, b) => b[1] - a[1])
  bi.sort((a, b) => b[1] - a[1])
  // Start with trigrams, then bigrams that are not contained in any kept trigram
  const keptTris: Str[] = []
  const keptBis: Str[] = []
  for (const [p] of tri) {
    if (!keptTris.includes(p)) keptTris.push(p)
  }
  for (const [p] of bi) {
    const lower = p.toLowerCase()
    const containedInTri = keptTris.some((tri) => tri.toLowerCase().includes(lower))
    if (!containedInTri) keptBis.push(p)
  }
  const phraseTokens = new Set<Str>()
  for (const p of keptTris.concat(keptBis)) for (const w of p.split(' ')) phraseTokens.add(w.toLowerCase())
  const filteredUni = uni.filter(([t]) => !phraseTokens.has(t.toLowerCase()))

  const merged: Array<[Str, number]> = [...tri, ...keptBis.map((p) => [p, 0] as [Str, number]), ...filteredUni]
  merged.sort((a, b) => b[1] - a[1])
  const out: Str[] = []
  const seen = new Set<Str>()
  for (const [term] of merged) {
    const key = term.toLowerCase()
    if (!seen.has(key)) { seen.add(key); out.push(term) }
    if (out.length >= topK) break
  }
  return out
}

if (require.main === module) {
  const text = process.argv.slice(2).join(' ') || 'Finasteride treats androgenetic alopecia by inhibiting types II and III 5α-reductase.'
  const kws = extractKeywords(text, 5)
  for (const k of kws) console.log(k)
}
